[client] 리뷰 요청

sangwon21 opened the pull request.

kiyoesjh committed:
[#30] fix: slider 오류 수정

why: storybook 에서 오류가 나서 Slider 에 있는 onInput 을 onChange 로 변경 했으며, Bar 에 key값을 줬다.

kiyoesjh committed:
[#30] feat: 메인 화면 구성을 위해 필요한 filters 추가

why: filters 폴더를 추가해 필터 관련 기능을 폴더 안으로 넣었으며, 그로 인해 경로 변경이 있었다. 메인 화면 구성을 위해 App.tsx 에서 컴포넌트 전체를 감싸고 가운데 정렬 할 수 있도록 수정했다.

sangwon21 committed:
[#33] feat: alias 추가

why: filters 폴더가 생겨서, alias를 추가했다.

sangwon21 committed:
[#33] feat: 필터링 데이터 전역으로 관리

why: 필터링 데이터를 지역적으로 관리하려고 하다보니, 중복된 코드가 많이
발생하고, 더불어, 필터링 데이터를 계층이 다른 컴포넌트에서 활용할 수
없어서, 전역으로 관리하기로 결정했다.

sangwon21 committed:
[#33] feat: card 데이터 저장을 위한 리듀서 작성

why: card 데이터를 서버에서 받아오면, 이를 저장할 공간이 필요했는데,
리덕스를 그 공간으로 활용하기로 결정했다. 이에 따라, 해당 작업을 수행할
수 있게 도와주는 리듀서를 작성했다.

sangwon21 committed:
[#33] feat: 비동기 통신을 위한 redux-saga 적용

why: 필터링 될 때마다, 해당 필터링 데이터를 이용하여, 서버에서 데이터를
비동기적으로 받아오려고, redux-saga를 이용했다. 비동기로 받아온 데이터는
다시 redux에 넣어줌으로써, 비동기 처리를 코드에서 진행할 필요가 없게
되었다. 하지만, react hooks를 잘 활용하면 redux-saga를 대체할 수 있을 것
같다. 어느 부분에 방향성을 둬야 될지 잘 모르겠다.

sangwon21 committed:
[#33] chore: redux-saga를 이용함에 따라 코드 변경

why: redux-saga를 사용함에 따라 코드의 변경이 필요하여, 변경했다.

sangwon21 committed:
[#33] feat: 가격 필터링 추가

why: 사용자가 가격을 기준으로 숙소 데이터를 조회할 수 있겠끔 하기
위하여, 가격을 바탕으로 필터링할 수 있는 기능을 추가하였다. 날짜
필터링과 인원 필터링을 기준으로 계속해서 가격 데이터를 바꿔야 하는
기능은 추후 구현 예정이다.

sangwon21 committed:
[#33] fix: 할인 가격 표시

why: 할인 가격을 숙소 데이터에 적시하기 위하여, 카드 컴포넌트를
수정했다.

sangwon21 committed:
[#33] feat: 헤더 작성

why: 프로젝트 내에서 필요한 헤더를 작성했다.

sangwon21 committed:
[#33] feat: 예약 버튼과 예약 모달 연결

why: 예약 버튼을 눌렀을 때, 해당 숙소 정보를 바탕으로 예약 모달이 나타날
수 있게끔, 버튼과 모달을 연결했다.

sangwon21 committed:
[#33] feat: 예약 버튼과 예약 모달 연결

why: 예약 버튼을 눌렀을 때, 해당 숙소 정보를 바탕으로 예약 모달이 나타날
수 있게끔, 버튼과 모달을 연결했다.

sangwon21 committed:
Merge branch 'client/feature/card-redux' of https://github.com/codesquad-member-2020/airbnb-13 into client/feature/card-redux

sangwon21 committed:
[#33] feat: 예약 확인 모달 추가

why: 서버에서 예약 확인 관련 데이터를 받아, 예약 확인 모달에 보여주는
작업을 수행했다. 더불어, 예약 버튼은, 날짜 필터링이나, 인원 필터링이
설정이 안 됐을 경우, 클릭할 수 없도록 막아놨다.

sangwon21 committed:
[#33] feat: Infinite Scroll 구현 중

crongro commented on client/src/reducer/cardReducer.ts:
```diff
@@ -0,0 +1,54 @@
+};
+
+const initialState = {
+  cardSet: 1,
```
cardSet : 1 
이라는 것이 어떤 것인지 유추하긴 어려워 보여요.
count인가 어떤 value인지 등을 판단하기 어려울 듯.

crongro commented on client/src/reducer/cardReducer.ts:
```diff
@@ -0,0 +1,54 @@
+    }
+    default:
+      return {
+        ...state
```
여기는 새로운 객체를 반환하지 않고 그대로의 값을 줘야 하겠죠.
shallow equality 할때도  '아 변경사항이 없구나' 라고 판단할 수 있겠죠.

crongro commented on client/src/reducer/filterReducer.ts:
```diff
@@ -0,0 +1,85 @@
+type FilterState = {
+  startDate: string;
+  endDate: string;
+  adult: number;
```
adultCount 라고 하는것도 좋을 듯

crongro commented on client/src/reducer/filterReducer.ts:
```diff
@@ -0,0 +1,85 @@
+const modalReducer = (state: FilterState = initialState, action: FilterAction) => {
+  switch (action.type) {
+    case SET_DATE_FILTER:
+      const { startDate, endDate } = action as ReturnType<typeof setDateFilter>;
```
오 ReturnType 활용을 한것이 멋져보이네요.

sangwon21 committed:
[#33] feat: 무한 스크롤링 구현

why: useEffect에서 Ref를 인식하지 못해서, React.Ref를 활용하지 않고,
대신 useState에서 직접 Element를 설정해주는 방식으로 선회했다.

crongro commented on client/src/saga/fetchCardData.ts:
```diff
@@ -0,0 +1,53 @@
+import axios from 'axios';
+
+function* fetchCardData() {
+  yield takeEvery([SET_DATE_FILTER, SET_GUEST_FILTER, SET_PRICE_FILTER], requestNewCard);
```
saga 적용 전체적으로잘하신거 같네요.

generator 공부를 꼭 같이 해주는 게 좋긴하죠!

takeEvery 와 takeLatest 와의 차이점 등도 공부하면 좋을 듯.

crongro commented on client/webpack.dev.js:
```diff
@@ -18,24 +18,32 @@ module.exports = {
   devtool: 'cheap-module-source-map',
```
디버깅 과정에서 꼭 활용하시길~

crongro commented on client/src/action/cardAction.ts:
```diff
@@ -0,0 +1,37 @@
+};
+
+export type CardAction =
+  | ReturnType<typeof setNewCardSet>
```
union타입 활용의 적절한 예군요.

crongro commented on client/src/components/Cards/Cards.tsx:
```diff
@@ -1,25 +1,72 @@
+  const dispatch = useDispatch();
+  useEffect(() => {
+    dispatch(setStartCardSet());
+  }, []);
```
한번 실행되는 코드인가보네요.
dependencies 로 넣어야 항목이 있는지 항상 확인해보세요.

crongro commented on client/src/components/Cards/Cards.tsx:
```diff
@@ -1,25 +1,72 @@
+      });
+    };
+
+    const io = new IntersectionObserver(handleIntersection, options);
```
IntersectionObserver ?? 오 활용한 이유 궁금하군요.

crongro commented on client/src/components/Filters/DateFilter/DateFilter.tsx:
```diff
@@ -5,13 +5,59 @@ import moment from 'moment';
+        startDateId="213242dfefsdve"
+        endDate={endDate}
+        endDateId="er32aesdfvder"
+        onDatesChange={({ startDate, endDate }) => {
```
property 도 많은데 함수까지 정의하니 jsx가 복잡해보이네요.
jsx는 늘 구조를 잘 드러내게 하는게 중요하니까 그런방향을 잘 유지해보세요.

crongro commented on client/src/components/Modal/Reservation/Prices/Prices.tsx:
```diff
@@ -2,15 +2,30 @@
+      <PriceRow category={'청소비'} price={cleaningFee}></PriceRow>
+      <PriceRow category={'서비스 수수료'} price={serviceFee}></PriceRow>
+      <PriceRow category={'숙박세와 수수료'} price={occupancyFee}></PriceRow>
+      <PriceRow category={'합계'} price={totalPrice}></PriceRow>
```
PriceRow 가 반복표현되어 있는데, category 값도 문자열로 jsx안에 그냥 들어가 있고요.

반복을 없애는 방법도 있다면 더 좋을거 같네요.

crongro commented on client/src/components/Modal/Reservation/Reservation.tsx:
```diff
@@ -6,14 +6,47 @@ import Header from './Header/Header';
 
 const Reservation = () => {
+  const { id } = useSelector((state: RootState) => state.modalReducer);
+  const [reservationInfo, setReservationInfo] = useState<ReservationInfo>({
```
컴포넌트 안에서만 필요한 값이면 이렇게 useState 로 상태관리하는게 좋겠죠.

crongro commented on client/src/components/Filters/GuestFilter/GuestFilter.tsx:
```diff
@@ -3,30 +3,27 @@ import { jsx, css } from '@emotion/core';
-        <Row type="성인" range={'만 13세 이상'} count={adult} setCount={setAdult} />
-        <Row type="어린이" range={'2~12세'} count={child} setCount={setChild} />
-        <Row type="유아" range={'2세 미만'} count={baby} setCount={setBaby} />
+        <Row type="성인" range={'만 13세 이상'} count={adult} setCount={(count: number) => dispatch(setGuestFilter("adult", count))} />
```
jsx안에서 이렇게 성인,어린이 이런 문자열도 뺄수 있다면 더 좋긴하죠.

crongro reviewed (approved):
잘 구현된 코드네요. 

프로젝트 디렉토리 구조도 계층적으로 잘 나눠서 좋고, 컴포넌트 크기들도 적당해보이네요.
jsx가 너무 복잡해지지 않는지? 검토만 좀더 하시면 될 듯.

브라우저호환성 이슈확인차, 
목표한 브라우저에서 잘 동작하는지 한번 확인해보세요.

saga 와 미들웨어에 대해서 잘 이해해보세요.

질문하나)
TS사용을 해보니 장점과 단점을 두 줄로 comment 남겨주세요.
@sangwon21 @kiyoesjh

kiyoesjh committed:
[#33] feat: 무한 스크롤링 구현

why: useEffect 에서 두번 실행되는 이유였던 observe 함수 호출을 삭제 해서 한번만 불리도록 만들었다. 관련 리듀서에서 SET_NEXT_CARD_SET 에서 cardSet을 올려 주던 것을 GET_NEXT_CARD_SET 에서 올려주도록 변경했다.

kiyoesjh committed:
[#33] refactor: 리뷰 반영

why: 가독성 측면에서 cardSet 을 cardPage 이름으로 변경했으며, state가 변경되지 않을 경우 새로 객체를 생성하지 않도록 수정했다.

kiyoesjh committed:
[#33] refactor: 필요 없는 부분 삭제

why: 필요없는 주석들을 삭제했다.
